### DISCUSSION TITLE
Redefine start/stop for local instances

### DISCUSSION THREAD

**Discussion by tailhook** [geldata/gel#3250] [Pre-RFC]
# Constraints

1. On WSL/windows, there is no systemd daemon.
2. It's also common to have no systemd on linux CI's (for github-actions we have the special action, but it's not easily possible for many other CIs).
3. On MacOS launchctl agents only work on GUI session, not on SSH by default.
4. Sometimes running `--foreground` is nice for debugging even on systems having process supervisor (systemd or launchctl).
5. Sometimes if you already run non-supervisor-supported instance, supervisor may appear later (i.e. gui session in MacOS started while ssh session is active).

# Specification

## Always use CLI

We change service defition from directly running server to supervise with CLI `edgedb instance xxx --from-systemd` (or `...--from-launchctl`. This does few jobs:

1. Allows updating server command-line easier
2. Prevents conflicts with server running using other way (`--foreground`, or just instance started before supervisor daemon started). This is done by checking pid file and maybe using advisory lock files.

I.e. if server is already running in foreground, you will see the following in the log after starting service:
```
EdgeDB Instance xxx is running in foreground as pid 123.
    Hint: Use `edgedb instance restart xxx` to kill the foreground instance and start as a service
```
Whereas `stop` and `restart` will stop even foreground instance and spawn as a service or daemonized.

## Daemonize manually

If CLI detects that there is no existing supervisor like systemd or launchctl it will daemonize itself, with the database running as the child process

This outputs the same state files and uses same advisory locks as any other invocation. So if launchctl GUI session appears suddenly it will emit the message described in previous section.

**Caveat:** we will still fail in `project init` and `instance create` unless `--start-conf=manual` is specified. Becase this way of daemonizing doesn't provide a good way for auto-start.

Note: for windows, we'll implement auto-start using this mode, but it will include some wrapper files outside of the WSL to do this on system startup.

## Nicer Foreground

If you want to debug your instance manually, we don't have to make you run stop manually as we required before, we can allow:
```
edgedb instance start xxx --foreground --auto-restart
```
It will stop the service from running (or kill the daemon), start the foreground server, and on exit of the foreground server it will restart the daemon/service automatically.

**Comment by elprans** [comment]
LGTM in general.

> `--from-systemd` (or `...--from-launchctl`)

Do we need to distinguish them explicitly?  Perhaps a generic `--from-service-manager` is sufficient?

> `edgedb instance start xxx --foreground --auto-restart`

Why not `edgedb instance restart xxx --foreground`?  Every service manager has a `restart` action, so that's what I'd reach for instinctively.

**Reply by tailhook** [reply]
> LGTM in general.
> 
> > `--from-systemd` (or `...--from-launchctl`)
> 
> Do we need to distinguish them explicitly? Perhaps a generic `--from-service-manager` is sufficient?
> 

Not sure. But maybe we need some other mode for running a service from alpine's init system. Anyway I think they will be hidden options, so it doesn't matter how they are named.

> > `edgedb instance start xxx --foreground --auto-restart`
> 
> Why not `edgedb instance restart xxx --foreground`? Every service manager has a `restart` action, so that's what I'd reach for instinctively.

It maybe not obvious that service will be restarted again after `Ctrl+C`. But generally yes, this is also an option.

**Reply by elprans** [reply]
Restarted in background?

**Reply by tailhook** [reply]
Yes. Kinda works as "docker attach" or "tmux attach" (except much slower, but with other upsides).
